Spec.md

```diff
@@ -0,0 +1,223 @@
+# OPC UA Pub/Sub Interface Description
+
+This document defines the Interface of the umati Dashboard based on OPC UA PubSub.
+It describes how a OPC UA server address space must be mapped to OPC UA PubSub, so that the infromation can map to the templates and displayed.
+
+## Requirements for the Mapping of the address space to PubSub
+
+(Reference: [umati/UA-CloudPublisher#5](https://github.com/umati/UA-CloudPublisher/issues/5))
+
+- Ability to react to dynamic changes in the address space (nodes added/deleted)
+  - Changes should be detected by monitoring NodeManagement events.
+- Resolve references to nodes (NodeId as values)
+- Hierarchy references must be mapped
+- Non-hierarchy references should be mapped if necessary
+  - Non-hierarchy references are mapped if relevant to the data semantics.
+- Provide type information
+- Methods should be considered
+- Online detection (possibly as separate topics)
+- Allow empty objects
+- Events/alarms must be mappable
+- Coordinate the completeness of the run
+- Encoding should be UA JSON
+- Dimensioning for approximately 350 subscribers (machines + dashboard users)
+- Custom DataTypes must be included
+- The address space includes at least OPC UA Machinery and possibly other Companion Specifications based on it.
+- The entry point of the mapping is the machine object in the Machines folder of Machinery.
+
+## Design Decisions
+
+The concept of OPC UA Part 14 should be followed as closely as possible.
+
+## Mapping
+
+### Mapping of Objects and Variables
+
+- An object is mapped to a DataSet.
+- A variable is mapped as a field to the DataSet. If the variable contains children (other variables), a DataSet with the children is created.
+
+#### Example
+
+![image](https://github.com/umati/UA-CloudPublisher/assets/12342602/91eae1d7-9d6f-4246-abb2-0ec5ea9b59bc)
+
+The image shows a part of an address space. Each object is mapped to a DataSet:
+
+- Production
+- ActiveProgram
+- State
+
+The properties Name, NumberInList, Id, and Number are fields of the DataSets. The value of CurrentState is mapped as a field of ActiveProgram and has a DataSet with its properties:
+
+- Production
+- ActiveProgram
+  - Name
+  - NumberInList
+  - CurrentState
+- State
+- CurrentState
+  - Id
+  - Number
+
+### Explanation
+
+A field can also contain properties but no other variables, even though the address space can contain more hierarchies. Properties of objects also cannot be mapped to the properties of a field. Therefore, a uniform mapping is used for objects and variables so that access is always consistent.
+
+## Mapping of References

```

Comment:
I think including the TypeDefintion of the Objects and maybe even the nested VariableTypes (e.g. FiniteStateVariableType for CurrentState) the is also needed, as the Frontend has some conditional rendering logic based on TypeDefinitions. I suggest adding Fields with TypeDefintion for the ObjectNode of the DataSet. E.g.: 
```json
        "Fields": [
            {
                "Name": "TypeDefinition",
                "Description": {
                    "Locale": "",
                    "Text": ""
                },
                "FieldFlags": 0,
                "BuiltInType": 0,
                "DataType": {
                    "Id": 0
                },
                "ValueRank": 1926237584,
                "ArrayDimensions": [],
                "MaxStringLength": 0,
                "DataSetFieldId": "57B13D4E-3F71-2A07-5AE5-1E9107E09EF9",
                "Properties": []
            },
```

Reply:
I´m not sure. The TypeDeifnition is given in the MetaData. If it is part of the DataSet a lot of value which do not change will send in every message.

Reply:
I don't think it matters much, except that using the description may not be semantically straightforward. 
Wouldn't we use Delta Frames if we were interested in low bandwidth anyway?

Reply:
Key and Delta Frames, should be followed.

Reply:
Arent the key-frames retained?

Reply:
Edit: DataSetMessages can be retained (Table 198 in Part 14). 

Where do you think using the field is violating the Spec? 

However, using the description as described above certainly does. The description is a LocalizedText and should be encoded as described here (https://reference.opcfoundation.org/Core/Part6/v105/docs/5.4.2.15). I.e., we would have use the there described JSON-ecoding of LocalizedText and encode it for example as escaped JSON string:
```json
...
"StructureDataTypes": "...",
"Name": "5:Production.5:ActiveProgram.5:State",
"Description": "{\r\n    \"Type\": \"5:ProductionProgramStateMachineType\",\r\n    \"TypeNodeID\": \"nsu=http:\/\/opcfoundation.org\/UA\/Machinery\/;i=58997\",\r\n    \"AdditionalReference\": [\r\n      {\"GeneratesEvents\":\"nsu=http:\/\/opcfoundation.org\/UA\/Machinery\/;i=3444\"}\r\n    ]\r\n  }",
...
```
Another problem with the description is, how to get the exact TypeDefinition (not DataType) for the nested VariableTypes. E.g, how to serialize the TypeDefintion (FiniteStateVariableType) of the CurrentState if it is part of the State's DataSet. 

Here IMO using the fields with retained messages for key frames would be more adaptable, as we could simply add multiple TypeDefinition fields for that use case. But of course we can also add that to the description, as long as we encode it correctly.

Reply:
Yes, but nothing in the spec forbids this, I guess, as arbitrary fields can be added. 
Ah ok, then this should be updated:
```
The image shows a part of an address space. Each object is mapped to a DataSet:

Production
ActiveProgram
State
```
If CurrentState gets its own message, we break the semantics, that each ObjectNode (object?) gets its own DataSet?
What would be the semantics? There is always an own DataSet for nodes, if they are object nodes or if they are VariableTypes with properties? It they are VariableTypes with properties they get their own Topic, otherwise they are added to the parent?

I find the simpler mapping of ObjectType -> Own topic, VariableType -> no own topic more straightforward.

Reply:
The semantic idea was that a topic contains the childs. So the value of a Variable is in the DataSet of the Object an the childs are in the new Topic. 

The question in a straightforward mapping is, how to handle that variable also can have childs and grandchilds?

Reply:
> Yes, but nothing in the spec forbids this, I guess, as arbitrary fields can be added.

Yes also nothing forbids to use the description. Both solution are ugly in my point of view. The description solution try to map metadata to metadata and data to data.

Reply:
@wlkrm I try to describe both variants. Can you please check if I describe the Field solution correct?
If it is okay for you I would suggest to send this doc with both variants to the OPC Foundation and get their opinion.